itest/lndsigner_test.go

```diff
@@ -0,0 +1,840 @@
+	idPubKey string
+
+	unixSocket bool
+
+	lndSignerCmd *exec.Cmd
+	lndSignerMtx sync.RWMutex
+
+	lndDir    string
+	lncliPath string
+	rpc       string
+	p2p       string
+	lndCmd    *exec.Cmd
+	lndMtx    sync.RWMutex
+}
+
+// Start takes the initial configuration (tctx, idPubKey, and unixSocket) and
+// starts lndsignerd and LND.
+func (l *lndHarness) Start() {
+	l.tctx.t.Helper()
+
+	// Start lndsignerd.
+	l.lndDir = path.Join(l.tctx.tmpRoot, fmt.Sprintf("lnd%s", l.idPubKey))
+	err := os.Mkdir(l.lndDir, fs.ModeDir|0700)
+	require.NoError(l.tctx.t, err)
+
+	signerAddr := "127.0.0.1:" + newPortString()
+	fullSignerAddr := "tcp://" + signerAddr
+
+	if l.unixSocket {
+		signerAddr := path.Join(l.lndDir, "signer.socket")
+		fullSignerAddr = "unix://" + signerAddr
+	}
+
+	l.lndSignerMtx.Lock()
+	defer l.lndSignerMtx.Unlock()
+
+	l.lndSignerCmd = exec.CommandContext(l.tctx.ctx, l.tctx.lndSignerPath,
+		"--rpclisten="+fullSignerAddr, "--nodepubkey="+l.idPubKey,
+		"--tlscertpath=./testdata/tls.cert",
+		"--tlskeypath=./testdata/tls.key", "--network=regtest",
+	)
+
+	l.lndSignerCmd.Env = append(l.lndSignerCmd.Env,
+		"VAULT_ADDR=http://127.0.0.1:"+l.tctx.vaultPort,
+		"VAULT_TOKEN=root",
+	)
+
+	l.lndSignerCmd.Stderr = bytes.NewBuffer(make([]byte, 0))
+
+	err = l.lndSignerCmd.Start()
+	require.NoError(l.tctx.t, err)
+	go waitProc(l.tctx.t, l.lndSignerCmd, l.lndSignerMtx)
+
+	// Start lnd.
+	acctsResp, err := l.tctx.vaultClient.ReadWithData(
+		"lndsigner/lnd-nodes/accounts",
+		map[string][]string{
+			"node": []string{l.idPubKey},
+		},
+	)
+	require.NoError(l.tctx.t, err)
+
+	acctList, ok := acctsResp.Data["acctList"].(string)
+	require.True(l.tctx.t, ok)
+
+	accounts, err := lndsigner.GetAccounts(acctList)
+	require.NoError(l.tctx.t, err)
+
+	grpcAccounts := make([]*itest.WatchOnlyAccount, 0,
+		len(accounts))
+
+	for derPath, xPub := range accounts {
+		grpcAccounts = append(grpcAccounts,
+			&itest.WatchOnlyAccount{
+				Purpose:  derPath[0],
+				CoinType: derPath[1],
+				Account:  derPath[2],
+				Xpub:     xPub,
+			})
+	}
+
+	l.rpc = newPortString()
+	l.p2p = newPortString()
+
+	l.lndMtx.Lock()
+	defer l.lndMtx.Unlock()
+
+	l.lndCmd = exec.CommandContext(l.tctx.ctx, l.tctx.lndPath,
+		"--lnddir="+l.lndDir, "--norest", "--listen="+l.p2p,
+		"--rpclisten="+l.rpc, "--trickledelay=1", "--bitcoin.active",
+		"--bitcoin.regtest", "--bitcoin.node=bitcoind",
+		"--bitcoind.rpcuser=user", "--bitcoind.rpcpass=password",
+		"--bitcoind.rpchost=127.0.0.1:"+l.tctx.bitcoinRPC,
+		"--bitcoind.zmqpubrawblock=tcp://"+l.tctx.bitcoinZB.String(),
+		"--bitcoind.zmqpubrawtx=tcp://"+l.tctx.bitcoinZT.String(),
+		"--remotesigner.enable",
+		"--remotesigner.rpchost="+signerAddr,
+		"--remotesigner.tlscertpath=./testdata/tls.cert",
+		"--remotesigner.macaroonpath=./testdata/signer.custom.macaroon",
+	)
+
+	l.lndCmd.Stderr = bytes.NewBuffer(make([]byte, 0))
+
+	err = l.lndCmd.Start()
+	require.NoError(l.tctx.t, err)
+	go waitProc(l.tctx.t, l.lndCmd, l.lndMtx)
```

Comment:
It's the same answer as https://github.com/bottlepay/lndsigner/pull/11#discussion_r1093742660 - since `waitProc` is a different goroutine from where `l.lndCmd` is created, we have to gate that access by a mutex or we get a race condition detected even if it's not really a race condition.

Reply:
I've removed the mutex and moved the goroutine management inside `waitProc` and fixed the race condition inside there.